Processing topology.updated in incorrect order can lead to incomplete graph

Hi,

I was investigating one of the errors we see in the end-to-end tests, more specifically. this one:

```
=================================== FAILURES ===================================
__________________ TestE2EMefEline.test_155_patch_evc_active ___________________

self = <tests.test_e2e_13_mef_eline.TestE2EMefEline object at 0x7efef8621280>

   def test_155_patch_evc_active(self):
       api_url = KYTOS_API + '/mef_eline/v2/evc/'
       evc1 = self.create_evc(100)

       payload = {
           "active": False,
       }

       # It sets a new circuit's creation_time
       response = requests.patch(api_url + evc1, data=json.dumps(payload),
                                 headers={'Content-type': 'application/json'})
       assert response.status_code == 400, response.text

       # It verifies EVC's data
       response = requests.get(api_url + evc1)
       data = response.json()
     assert data['active'] is True
E       assert False is True

tests/test_e2e_13_mef_eline.py:941: AssertionError
=============================== warnings summary ===============================
tests/test_e2e_13_mef_eline.py::TestE2EMefEline::test_155_patch_evc_active: 2022-11-17,05:24:15.232687 - 2022-11-17,05:24:15.575109

```


and I found the following situation (logs filtered to only show events of interest):

```
Nov 17 05:24:15 runner-rmtk68xb-project-107-concurrent-0 2022-11-17 05:24:15,517 (Thread-106) kytos.napps.kytos/mef_eline:WARNING evc:659:  EVC(b6c744f7c52544, Vlan_100) was not deployed. No available path was found.
Nov 17 05:24:15 runner-rmtk68xb-project-107-concurrent-0 2022-11-17 05:24:15,519 (Thread-106) kytos.napps.kytos/mef_eline:DEBUG main:242:  create_circuit result {'circuit_id': 'b6c744f7c52544'} 201
Nov 17 05:24:15 runner-rmtk68xb-project-107-concurrent-0 2022-11-17 05:24:15,525 (Thread-106) werkzeug:INFO _internal:225:  127.0.0.1 - - [17/Nov/2022 05:24:15] "#033[35m#033[1mPOST /api/kytos/mef_eline/v2/evc/ HTTP/1.1#033[0m" 201 -
Nov 17 05:24:15 runner-rmtk68xb-project-107-concurrent-0 2022-11-17 05:24:15,545 (Thread-112) kytos.napps.kytos/mef_eline:DEBUG main:252:  update /v2/evc/b6c744f7c52544
Nov 17 05:24:15 runner-rmtk68xb-project-107-concurrent-0 2022-11-17 05:24:15,546 (Thread-112) kytos.napps.kytos/mef_eline:ERROR main:281:  active can't be updated.
Nov 17 05:24:15 runner-rmtk68xb-project-107-concurrent-0 2022-11-17 05:24:15,546 (Thread-112) kytos.napps.kytos/mef_eline:DEBUG main:282:  update result active can't be updated. 400
Nov 17 05:24:15 runner-rmtk68xb-project-107-concurrent-0 2022-11-17 05:24:15,548 (Thread-112) werkzeug:INFO _internal:225:  127.0.0.1 - - [17/Nov/2022 05:24:15] "#033[31m#033[1mPATCH /api/kytos/mef_eline/v2/evc/b6c744f7c52544 HTTP/1.1#033[0m" 400 -
Nov 17 05:24:15 runner-rmtk68xb-project-107-concurrent-0 2022-11-17 05:24:15,565 (Thread-113) kytos.napps.kytos/mef_eline:DEBUG main:137:  get_circuit /v2/evc/b6c744f7c52544
Nov 17 05:24:15 runner-rmtk68xb-project-107-concurrent-0 2022-11-17 05:24:15,569 (Thread-113) kytos.napps.kytos/mef_eline:DEBUG main:144:  get_circuit result {'active': False, 'archived': False, 'backup_path': [], 'bandwidth': 0, 'circuit_scheduler': [], 'current_path': [], 'dynamic_backup_path': True, 'enabled': True, 'failover_path': [], 'id': 'b6c744f7c52544', 'metadata': {}, 'name': 'Vlan_100', 'primary_path': [], 'service_level': 0, 'uni_a': {'tag': {'tag_type': 1, 'value': 100}, 'interface_id': '00:00:00:00:00:00:00:01:1'}, 'uni_z': {'tag': {'tag_type': 1, 'value': 100}, 'interface_id': '00:00:00:00:00:00:00:02:1'}, 'sb_priority': None, 'owner': None, 'queue_id': None, 'primary_constraints': {'spf_attribute': 'hop'}, 'secondary_constraints': {'spf_attribute': 'hop'}, 'primary_links': [], 'backup_links': [], 'start_date': '2022-11-17T05:24:15', 'creation_time': '2022-11-17T05:24:15', 'request_time': '2022-11-17T05:24:15', 'end_date': None} 200
Nov 17 05:24:15 runner-rmtk68xb-project-107-concurrent-0 2022-11-17 05:24:15,570 (Thread-113) werkzeug:INFO _internal:225:  127.0.0.1 - - [17/Nov/2022 05:24:15] "GET /api/kytos/mef_eline/v2/evc/b6c744f7c52544 HTTP/1.1" 200 -
```

Investigating a little further on the logs to see previous events, I found these log entries with seems to be the source of the problem:

```
Nov 17 05:24:03 runner-rmtk68xb-project-107-concurrent-0 2022-11-17 05:24:03,264 (thread_pool_app_5) kytos.napps.kytos/pathfinder:DEBUG main:243:  Topology graph updated with switches: ['00:00:00:00:00:00:00:01', '00:00:00:00:00:00:00:03', '00:00:00:00:00:00:00:02'], links: [].
Nov 17 05:24:03 runner-rmtk68xb-project-107-concurrent-0 2022-11-17 05:24:03,443 (thread_pool_app_6) kytos.napps.kytos/pathfinder:DEBUG main:243:  Topology graph updated with switches: ['00:00:00:00:00:00:00:01', '00:00:00:00:00:00:00:03', '00:00:00:00:00:00:00:02'], links: ['c8b55359990f89a5849813dc348d30e9e1f991bad1dcb7f82112bd35429d9b07', '4d42dc0852278accac7d9df15418f6d921db160b13d674029a87cef1b5f67f30'].
Nov 17 05:24:03 runner-rmtk68xb-project-107-concurrent-0 2022-11-17 05:24:03,464 (thread_pool_app_8) kytos.napps.kytos/pathfinder:DEBUG main:243:  Topology graph updated with switches: ['00:00:00:00:00:00:00:01', '00:00:00:00:00:00:00:03', '00:00:00:00:00:00:00:02'], links: ['c8b55359990f89a5849813dc348d30e9e1f991bad1dcb7f82112bd35429d9b07', '4d42dc0852278accac7d9df15418f6d921db160b13d674029a87cef1b5f67f30', '78282c4d5b579265f04ebadc4405ca1b49628eb1d684bb45e5d0607fa8b713d0'].
Nov 17 05:24:03 runner-rmtk68xb-project-107-concurrent-0 2022-11-17 05:24:03,468 (thread_pool_app_4) kytos.napps.kytos/pathfinder:DEBUG main:243:  Topology graph updated with switches: ['00:00:00:00:00:00:00:01', '00:00:00:00:00:00:00:03', '00:00:00:00:00:00:00:02'], links: ['c8b55359990f89a5849813dc348d30e9e1f991bad1dcb7f82112bd35429d9b07'].
Nov 17 05:24:15 runner-rmtk68xb-project-107-concurrent-0 2022-11-17 05:24:15,511 (Thread-111) kytos.napps.kytos/pathfinder:DEBUG main:188:  POST v2/ payload data: {'source': '00:00:00:00:00:00:00:01:1', 'destination': '00:00:00:00:00:00:00:02:1', 'spf_max_paths': 2, 'spf_attribute': 'hop', 'mandatory_metrics': {}, 'flexible_metrics': {}, 'minimum_flexible_hits': None}
Nov 17 05:24:15 runner-rmtk68xb-project-107-concurrent-0 2022-11-17 05:24:15,512 (Thread-111) kytos.napps.kytos/pathfinder:DEBUG main:214:  Found paths: []
Nov 17 05:24:15 runner-rmtk68xb-project-107-concurrent-0 2022-11-17 05:24:15,512 (Thread-111) kytos.napps.kytos/pathfinder:DEBUG main:221:  Filtered paths: []
```

From the logs above we can see that the pathfinder processed the `topology.updated` Events in a wrong order, which led to an incomplete graph and then fail to find a path. The graph became incomplete because the last `topology.updated` event processed (at `2022-11-17 05:24:03,468`) said that there is only one link in the topology, and considering the way pathfinder updates the graph (see [1]) it will remove all other links.

[1] https://github.com/kytos-ng/pathfinder/blob/5d9c9d0e75bd2b969a2f63a9238bfcd34b28c894/graph.py#L46

Analyzing this scenario, I guess what happened was: the links went up at the very same moment, triggering multiple `topology.updated` events, and then a wrong execution order led pathfinder to process those events incorrectly. Further analyzing pathfinder showed that the Lock mechanism used to avoid race conditions might be the reason for the wrong execution order: once many threads get locked, there is no guarantee on the order they will execute:

https://github.com/kytos-ng/pathfinder/blob/5d9c9d0e75bd2b969a2f63a9238bfcd34b28c894/main.py#L238-L240

With that hypothesis in mind, I tried to reproduce the behavior without success. I repeated the test with more than 400 executions (see script used in [2]), and all executions returned the correct path. After some investigation and some tries, I realized the environment where end-to-end tests are executed has a certain number of vCPUs. Then I executed again with the same number of vCPUs, and I was able to reproduce the issue. Once I reproduced the behavior, I added a versioning system to the `topology.updated` event and logged that version in the pathfinder processing routine. This showed me that all the times such error happened, it was due to processing the events in the wrong order. Out of 150 executions, 5 came out with errors.

In fact, since the `topology.updated` event provides information about the whole new topology on each event, if we missed one event and processed the newest one, there is no impact on just discarding the old one.

With that in mind, I created the following workaround: the topology will tag the `topology.updated` messages with a versioning system, then pathfinder will discard out-of-order events when processing them. The following proof-of-concept was implemented:

```
root@a9ea0d07cf66:/src/kytos-topology# git diff
diff --git a/main.py b/main.py
index f2a7d8a..341acf0 100644
--- a/main.py
+++ b/main.py
@@ -235,7 +235,8 @@ class Main(KytosNApp):  # pylint: disable=too-many-public-methods
             content={
                 'topology': self._get_topology(),
                 'failed_switches': failed_switches,
-                'failed_links': failed_links
+                'failed_links': failed_links,
+                'version': time.time(),
             })
         self.controller.buffers.app.put(event)

@@ -542,7 +543,8 @@ class Main(KytosNApp):  # pylint: disable=too-many-public-methods
         """Notify current topology."""
         name = "kytos/topology.current"
         event = KytosEvent(name=name, content={"topology":
-                                               self._get_topology()})
+                                               self._get_topology(),
+                                               "version": time.time()})
         self.controller.buffers.app.put(event)

     @listen_to("kytos/topology.get")
@@ -932,7 +934,8 @@ class Main(KytosNApp):  # pylint: disable=too-many-public-methods
         """Send an event to notify about updates on the topology."""
         name = 'kytos/topology.updated'
         event = KytosEvent(name=name, content={'topology':
-                                               self._get_topology()})
+                                               self._get_topology(),
+                                               'version': time.time()})
         self.controller.buffers.app.put(event)

     def notify_link_status_change(self, link, reason='not given'):




root@a9ea0d07cf66:/src/kytos-pathfinder# git diff
diff --git a/main.py b/main.py
index d98c7eb..53422ac 100644
--- a/main.py
+++ b/main.py
@@ -22,6 +22,7 @@ class Main(KytosNApp):
         """Create a graph to handle the nodes and edges."""
         self.graph = KytosGraph()
         self._topology = None
+        self._topology_version = 0
         self._lock = Lock()

     def execute(self):
@@ -235,12 +236,16 @@ class Main(KytosNApp):
         if "topology" not in event.content:
             return
         topology = event.content["topology"]
+        version = event.content.get("version", 0)
         with self._lock:
+            if version < self._topology_version:
+                return
             self._topology = topology
             self.graph.update_topology(topology)
+            self._topology_version = version
         switches = list(topology.switches.keys())
         links = list(topology.links.keys())
-        log.debug(f"Topology graph updated with switches: {switches}, links: {links}.")
+        log.debug(f"Topology graph updated with switches: version: {version}, switches: {switches}, links: {links}.")

     def update_links_metadata_changed(self, event) -> None:
         """Update the graph when links' metadata are added or removed."""
```

After applying the above workaround, I tested again, and out of 400 executions, no error was reported.

What do you guys think about this approach? @viniarck 

[2] Script used to reproduce the issue (inside the amlight/kytos:latest docker container)

```
# git clone https://github.com/amlight/kytos-end-to-end-tests
# cd kytos-end-to-end-tests
# export TESTS=tests/test_e2e_13_mef_eline.py::TestE2EMefEline::test_155_patch_evc_active
# ./kytos-init.sh
# sed -i s/INFO/DEBUG/g /etc/kytos/logging.ini
# cat bla.py
import requests
import time
from tests.helpers import NetworkTest

CONTROLLER = '127.0.0.1'
KYTOS_API = 'http://%s:8181/api/kytos' % CONTROLLER

net = NetworkTest(CONTROLLER)
net.start()

def setup_method():
    net.config_all_links_up()
    net.start_controller(clean_config=True, enable_all=True)
    net.wait_switches_connect()
    time.sleep(7)

def run_test(i=None):
    t1 = time.time()
    setup_method()
    api_url = KYTOS_API + '/pathfinder/v2/'
    a = {'source': '00:00:00:00:00:00:00:01:1', 'destination': '00:00:00:00:00:00:00:03:1', 'spf_max_paths': 10, 'spf_attribute': 'hop', 'mandatory_metrics': {}, 'flexible_metrics': {}, 'minimum_flexible_hits': None}
    response = requests.post(api_url, json=a)
    t2 = time.time()
    print(response.status_code)
    print(response.json())
    print(i, t2 - t1)


for i in range(400):
    run_test(i)

net.stop()

# python3 bla.py
```

--- a/CHANGELOG.rst
+++ b/CHANGELOG.rst
@@ -23,6 +23,7 @@ Removed
 Fixed
 =====
 - Fixed ``undesired_links`` (logical OR) and ``desired_links`` (logical AND) filters.
+- Safe guards to discard graph topology older (out of order) events
 
 Security
 ========

--- a/main.py
+++ b/main.py
@@ -23,6 +23,8 @@ class Main(KytosNApp):
         self.graph = KytosGraph()
         self._topology = None
         self._lock = Lock()
+        self._topology_updated_at = None
+        self._links_updated_at = {}
 
     def execute(self):
         """Do nothing."""
@@ -236,7 +238,13 @@ class Main(KytosNApp):
             return
         topology = event.content["topology"]
         with self._lock:
+            if (
+                self._topology_updated_at
+                and self._topology_updated_at > event.timestamp
+            ):
+                return
             self._topology = topology
+            self._topology_updated_at = event.timestamp
             self.graph.update_topology(topology)
         switches = list(topology.switches.keys())
         links = list(topology.links.keys())
@@ -247,7 +255,13 @@ class Main(KytosNApp):
         link = event.content["link"]
         try:
             with self._lock:
+                if (
+                    link.id in self._links_updated_at
+                    and self._links_updated_at[link.id] > event.timestamp
+                ):
+                    return
                 self.graph.update_link_metadata(link)
+                self._links_updated_at[link.id] = event.timestamp
             metadata = event.content["metadata"]
             log.debug(f"Topology graph updated link id: {link.id} metadata: {metadata}")
         except KeyError as exc:

--- a/tests/unit/test_main.py
+++ b/tests/unit/test_main.py
@@ -2,6 +2,7 @@
 
 from unittest import TestCase
 from unittest.mock import MagicMock, patch
+from datetime import timedelta
 
 from kytos.core.events import KytosEvent
 from kytos.lib.helpers import get_controller_mock, get_test_client
@@ -28,6 +29,29 @@ class TestMain(TestCase):
         self.napp.update_topology(event)
 
         self.assertEqual(self.napp._topology, topology)
+
+    def test_update_topology_events_out_of_order(self):
+        """Test update topology events out of order.
+
+        If a subsequent older event is sent, then the topology
+        shouldn't get updated.
+        """
+        topology = get_topology_mock()
+        assert self.napp._topology_updated_at is None
+        first_event = KytosEvent(
+            name="kytos.topology.updated", content={"topology": topology}
+        )
+        self.napp.update_topology(first_event)
+        assert self.napp._topology_updated_at == first_event.timestamp
+        assert self.napp._topology == topology
+
+        second_topology = get_topology_mock()
+        second_event = KytosEvent(
+            name="kytos.topology.updated", content={"topology": second_topology}
+        )
+        second_event.timestamp = first_event.timestamp - timedelta(seconds=10)
+        self.napp.update_topology(second_event)
+        assert self.napp._topology == topology
 
     def test_update_topology_failure_case(self):
         """Test update topology method to failure case."""
@@ -269,6 +293,31 @@ class TestMain(TestCase):
         self.napp.update_links_metadata_changed(event)
         assert self.napp.graph.update_link_metadata.call_count == 1
         assert self.napp.controller.buffers.app.put.call_count == 0
+
+    def test_update_links_changed_out_of_order(self):
+        """Test update_links_metadata_changed out of order."""
+        self.napp.graph.update_link_metadata = MagicMock()
+        self.napp.controller.buffers.app.put = MagicMock()
+        link = MagicMock(id="1")
+        assert link.id not in self.napp._links_updated_at
+        event = KytosEvent(
+            name="kytos.topology.links.metadata.added",
+            content={"link": link, "metadata": {}}
+        )
+        self.napp.update_links_metadata_changed(event)
+        assert self.napp.graph.update_link_metadata.call_count == 1
+        assert self.napp.controller.buffers.app.put.call_count == 0
+        assert self.napp._links_updated_at[link.id] == event.timestamp
+
+        second_event = KytosEvent(
+            name="kytos.topology.links.metadata.added",
+            content={"link": link, "metadata": {}}
+        )
+        second_event.timestamp = event.timestamp - timedelta(seconds=10)
+        self.napp.update_links_metadata_changed(second_event)
+        assert self.napp.graph.update_link_metadata.call_count == 1
+        assert self.napp.controller.buffers.app.put.call_count == 0
+        assert self.napp._links_updated_at[link.id] == event.timestamp
 
     def test_update_links_changed_key_error(self):
         """Test update_links_metadata_changed key_error."""

